### PR TITLE
Migrate setup.py to pyproject.toml

### PR DESCRIPTION
**Pull request by kyluca**
Implements the first half of the migration as discussed on #738.

Essentially followed the official advice from https://packaging.python.org/en/latest/guides/modernize-setup-py-project/. Every field from the old `setup.py` config has an analogue in `pyproject.toml` just with slightly different syntax.

I've also confirmed that the project can still be built into both a source distribution (sdist) and a wheel (bdist). Corresponding documentation has been updated from using `python setup.py ... sdist bdist_wheel` to the newer `python -m build` as the file `setup.py` itself is no longer required.

This is a screenshot of the contents of the wheel of the latest release downloaded from PyPI on the left hand side vs. the contents of a locally built wheel for a 0.103 rc1 package showing the non `*.py` package data files are still present.

screenshot_from_2024_10_12_00_17_43

I've also verified the unit test suite still passes locally after all the changes (in particular to how things like package data is handled).

Let me know if there's anything else I can update at this stage. If this looks okay I'll continue with a follow up PR on consolidating all of the `pip`, `pip-tools`, `pipx` and `venv` usage to just `uv` (second half of the plan as discussed on the issue). Although that second set of changes is not required, it's just a "nice to have".

### CODE REVIEW

**Comment by akrabat** [comment]
Thanks!

Two quick questions:

* Do we need to update line 49 in tox.ini? (Though I suspect tox.ini needs looking at in its entirety…)
* We reference `setup.py` in the manual. I think it's benign as it is in an example code block, but maybe we should use a different example?

**Comment by akrabat** [comment]
Note that we need @lornajane to review my changes to https://github.com/rst2pdf/rst2pdf/pull/1236 and get them into main before we rebase/pull them into this branch to get the CI to run.

File: doc/RELEASE_PROCESS.rst [hunk 1]
```diff
@@ -72,11 +72,15 @@ This is an outline of what needs to be done in order to release rst2pdf.
...
     them as otherwise uploading to PyPI will not work.
 
+    Note this step uses a workaround as per https://github.com/pypa/setuptools/issues/4129 as ``build`` and
+    ``setuptools`` don't have full support for appending arbitrary tags to version numbers at build time.
+
     ::
 
-       $ python setup.py egg_info -b "rc1" sdist bdist_wheel
+       $ echo -e "[egg_info]\ntag_build=rc1\n" > /tmp/build_opts.cfg
+       $ DIST_EXTRA_CONFIG=/tmp/build_opts.cfg python -m build
```
Comment (akrabat) [inline doc/RELEASE_PROCESS.rst:81]: Need an  `rm` for the tmp file created to keep things tidy.

**Review by akrabat — COMMENTED** [review]
Can you update the `pip install` command in `RELEASE_PROCESS.rst` to match the update in `DEVELOPERS.rst` and `CONTRIBUTING.rst`, please.

**Comment by kyluca** [comment]
Thanks for the review 😄 

> Do we need to update line 49 in tox.ini? (Though I suspect tox.ini needs looking at in its entirety…)

Lorna mentioned in the issue that `tox.ini` is no longer used, but I wanted to try and keep this first PR as simple as possible. Happy to split that removal into a follow up PR that's scoped to `tox` alone, or bundle it in with my PR to introduce `uv`. Either way, I've updated L49 in the meantime.

> We reference setup.py in the manual. I think it's benign as it is in an example code block, but maybe we should use a different example?

Done ✅ 

> Note that we need @lornajane to review my changes to https://github.com/rst2pdf/rst2pdf/pull/1236 and get them into main before we rebase/pull them into this branch to get the CI to run.

No worries, that makes sense

> Can you update the pip install command in RELEASE_PROCESS.rst to match the update in DEVELOPERS.rst and CONTRIBUTING.rst, please.

Done ✅ 

> Need an rm for the tmp file created to keep things tidy.

Done ✅

**Comment by lornajane** [comment]
That dependent PR is now merged, and I think the build failures here will be fixed if you update the branch

**Comment by kyluca** [comment]
Thanks, rebased and pushed now. Ready to test 👍🏾

**Comment by akrabat** [comment]
> > Do we need to update line 49 in tox.ini? (Though I suspect tox.ini needs looking at in its entirety…)
> 
> Lorna mentioned in the issue that `tox.ini` is no longer used, but I wanted to try and keep this first PR as simple as possible. Happy to split that removal into a follow up PR that's scoped to `tox` alone, or bundle it in with my PR to introduce `uv`. Either way, I've updated L49 in the meantime.

I missed that. Yeah, we can deal with that in a separate PR.

**Review by akrabat — APPROVED** [review]
LGTM

**Pull request merged**